[SPBU-21.B10] Task 2.2 Мясникова Мария

mawekk opened the pull request.

mawekk committed:
hw2.2 (not done)

mawekk committed:
Merge branch 'master' into hw2_task_2

mawekk committed:
hw2.2 (not done)

mawekk committed:
Merge remote-tracking branch 'origin/hw2_task_2' into hw2_task_2

mawekk committed:
hw2.2 (not done)

mawekk committed:
hw2_task2 not done

mawekk committed:
styleguide

ArtyomLobanov commented on homework_2/task_2.c:
```diff
@@ -0,0 +1,58 @@
+void readFile(FILE* fileName)
+{
+    List* list = makeNewList();
+    int size = 0, numberOfLogs = 0;
```
Формально говоря, используемый нами стайлгайд такое не запрещает, но вообще объявлять несколько переменных на одной строке – дурной тон. Так тяжелее следить за количеством и изменением переменных + конкретно в C/C++ есть проблема с указателями. В конструкции
```cpp
int* a, b, c;
```
* на самом деле относится к имени переменной т.е. вы объявили один указатель на int и две переменные типа int. И это выглядит контринтуитивно и может сбивать с толку при беглом чтении. Объявить три указателя можно было бы так:
```cpp
int *a, *b, *c;
```
Но лучше в целом избегать таких конструкций. Экономить строки никто не просил.

ArtyomLobanov commented on homework_2/task_2.c:
```diff
@@ -0,0 +1,58 @@
+    List* list = makeNewList();
+    int size = 0, numberOfLogs = 0;
+    fscanf(fileName, "%d", &size);
+    char originalData[size], log[128] = "", firstData[128] = "", secondData[128] = "";
```
Сколько нужно места, чтобы хранить строку из size символов? Спойлер: size + 1 т.к. строка должна оканчиваться специальным нулевым символом.

ArtyomLobanov commented on homework_2/task_2.c:
```diff
@@ -0,0 +1,58 @@
+    return (fileName != NULL);
+}
+
+void readFile(FILE* fileName)
```
FILE* – это файловый дескриптор. Можно было обозвать его file, input и т.д., но имя filename этому типу точно не соответствует.

ArtyomLobanov commented on library/DNA.c:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    printf("del: ");
+    makeNewElement(map, dna);
```
У вас утечка памяти: вы создаёте копию dna здесь, а потом создаёте ещё копию копии в makeNewElement, а память от первой копии не очищаете. Давайте примем за правило, что в makeNewElement вызывается только изнутри структуры и гарантируется, что переданный указатель можно использовать не копируя

ArtyomLobanov commented on library/DNA.c:
```diff
@@ -0,0 +1,86 @@
+    for (int i = 0; i < strlen(first); ++i) {
+        for (int j = 0; j < strlen(dna); ++j)
+            first[j] = first[j + 1];
+    }
```
Тут что-то странное написано. Во-первых, удалить часть строки можно за линейное время, а не за квадратное, как у вас. Во-вторых, first указывает куда-то в середину строки. Когда вы обращаетесь к first[strlen(dna) - 1], вы попадаете куда-то далеко за пределы массива. Ну и strlen работает не за 1, а за длинну строки. Т.е. на каждой итерации for-а при проверке условия вы итерируетесь по всей строке.

ArtyomLobanov commented on library/DNA.c:
```diff
@@ -0,0 +1,86 @@
+    }
+    for (int i = 0; i < strlen(end); ++i) {
+        for (int j = 0; j < strlen(dna); ++j)
+            first[j] = first[j + 1];
```
А позицию end вы вообще не искали же? Что здесь происходит?

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+struct ListElement {
+    char symbol;
+    ListElement* nextElement;
+    ListElement* previousElement;
```
Так, а DNA.c – это кусок какого-то альтернативного решения, заброшенного на полпути да? Удалите его из репозитория тогда

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        list->head = element;
+        element->previousElement = NULL;
+        element->nextElement = list->tail;
+    } else if (!list->tail) {
```
Не правда ли `list->head == NULL` и `list->tail == NULL` случаются только в одном случае (и всегда одновременно) – если список пуст? Тогда нужен один if для пустого списка и один для непустого. А сейчас у вас бага – при добавлении элемента в пустой список не обновляется указатель на хвост.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        }
+        if (!(strcmp(string, fragment)))
+            break;
+    }
```
Давайте в случае неудачи либо записывать NULL в beginning и ending, либо возвращать false из findFragment.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        for (ListElement* element = current; element; element = element->nextElement) {
+            string[i] = element->symbol;
+            i++;
+            if (strlen(string) == strlen(fragment)) {
```
strlen итерируется по строке, пока не найдёт нулевой символ. Это долгая операция. Не вызывайте её без острой необходимости. Определить тот факт, что мы только что скопировали последний символ можно проще `if (!fragment[i + 1])`.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+                break;
+            }
+        }
+        if (!(strcmp(string, fragment)))
```
Так, конечно, можно, но это требует выделения дополнительной памяти. Ну и строго говоря, ограничение в 128 символов относится к нашей конкретной задаче, а эта реализация списка символов лежит в папке абсолютно справедливо лежит в папке library и, следовательно, должна быть универсальной. Давайте реализуем функцию match:
```cpp
bool match(Element* start, char* pattern)
{
    Element* current = start;
    char* pointer = pattern;
    while (*pointer && current) {
        if (*pointer != current->symbol) 
            return false;
        pointer++;
        current = current->next;
    }
    return !(*pointer);
}
```
(Конкретно на такой реализации не настаиваю, главное что без использования дополнительной памяти и встроенных функций сравнения строк)

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+    }
+}
+
+void findFragment(List* list, ListElement** beginning, ListElement** ending, char* fragment)
```
Но вам же нужно будет искать фрагмент не только с начала списка, но и с какого-то места в нём. Скажем, при выполнении команды `DELETE AB CD` для строки `CDABAACDAB` вам нужно будет сначала найти первый фрагмент AB, а затем первый фрагмент `CD`, начинающийся строго после конца найденного AB. Т.е. как-то так `CD[AB]AA[CD]AB`, а не так `[CD][AB]AACDAB`

ArtyomLobanov replied:
Поэтому принимать здесь нужно не указатель на list целиком, а указатель на конкретный элемент списка, с которого следует начать поиск.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+
+void insertElements(ListElement* current, char* fragment)
+{
+    for (int i = 0; i < strlen(fragment); ++i) {
```
Опять у вас strlen в условии цикла используется :(
```cpp
for (char* pointer = fragment; *pointer; pointer++) { // Пока не встретили нулевой символ = конец строки
    ... // добавляем в список символ, на который указывает pointer
}
```

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        new->previousElement = current;
+        new->nextElement = current->nextElement;
+        current->nextElement = new;
+        current = new;
```
Если current – последний элемент списка, то нужно обновить list->tail

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        next->previousElement = element->previousElement;
+        free(element);
+        element = next;
+        next = element->nextElement;
```
Если start – первый элемент списка или end – последний, то нужно обновить соответствующие поля в list

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+    ListElement* endBeginning = NULL;
+    ListElement* endEnding = NULL;
+    findFragment(list, &startBeginning, &startEnding, start);
+    findFragment(list, &endBeginning, &endEnding, end);
```
Вы сейчас можете найти конец в строке раньше, чем начало. Пример: `DELETE AB CD` для строки `CDABAACD`

ArtyomLobanov replied:
И проверяйте, что фрагменты в принципе нашлись. Даже если вы верите в прямоту моих рук (зря) и уверены, что в тестовом файле нет ошибок, дополнительные проверки всегда могут помочь найти и локализовать баги в вашем коде.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+        next = element->nextElement;
+        free(element);
+    }
+}
```
Забыли освободить память из под самого списка (List)

ArtyomLobanov reviewed (changes requested):
На исправление неделя. Уложитесь – поставлю полный балл

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,152 @@
+    ListElement* current = templateBeginning->previousElement;
+    deleteElements(templateBeginning, templateEnding);
+
+    insertElements(current, fragment);
```
Кстати говоря, если мы делаем replace от начала строки (`REPLACE ABA CABA` для `ABAFCYGVUIO`), то current у вас будет NULL и insertElements скорее всего упадёт.

ArtyomLobanov commented:
Это же вы жаловались на replace? Скорее всего проблема в необновляемых ссылках на голову списка. Если всё исправите, а оно не заработает, пните ещё раз

mawekk committed:
Merge branch 'master' into hw2_task_2

mawekk committed:
delete DNA.c

mawekk committed:
delete DNA.h

mawekk committed:
наелся и уснул

mawekk committed:
Merge remote-tracking branch 'origin/hw2_task_2' into hw2_task_2

mawekk committed:
наелся и уснул из-за пустой строки в коде

mawekk commented:
пыталась чего-то исправлять, но захотела потестить вывод в файл и, конечно же, все сломалось👍🏻
почему-то выводит 18 строк (хотя операций делает штук 30) и кусок 19 строки :)
что не так - вообще не понимаю (плакать хочется)

ArtyomLobanov commented:
Если не ошибаюсь, запись в файл через стандартный fprintf буферизируется т.е. когда вы вызываете fprintf, ваши символы записываются не сразу (запись в файл – долгая операция), а сначала добавляются в буфер в оперативной памяти. Когда буфер заполняется, данные сразу скопом переписываются в файл. Но если программа упадёт раньше, то данные в буфере пропадут. Меньше строк в файле у вас получается скорее всего именно по этой причине. А код сейчас гляну

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+        list->head = element;
+        element->previousElement = NULL;
+        list->tail = element;
+        list->tail->nextElement = NULL;
```
element->previousElement  и element->nextElement и так равны NULL при создании элемента через makeNewElement

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+{
+    ListElement* element = start;
+    ListElement* next = element->nextElement;
+    while (element != end->nextElement) {
```
Вот тут есть проблема. Вот вы дошли до element == end. В цикле вы освобождаете память из-под этого элемента и следующее значение в element. Затем проверяется условие element != end->nextElement, но вот беда: т.к. память на которую ссылается end мы только что освободили, теперь мы читаем данные из освобождённого участка памяти. Это Undefined Behaviour. Конкретно на моём пк ваша программа падала при выполнении первой команды. Если добавить дебаговый вывод, то выглядит это так:
```cpp
void deleteElements(ListElement* start, ListElement* end, List* list)
{
    ListElement* element = start;
    ListElement* next = element->nextElement;
    printf("\n%d\n", end->nextElement);
    while (element != end->nextElement) {
        printf("%d %d %d %c\n", element, end->nextElement, element != end->nextElement, element->symbol);
        element->previousElement->nextElement = next;
        next->previousElement = element->previousElement;
        free(element);
        element = next;
        next = element->nextElement;
    }
```
Вывод:
```
16848528
16845576 16848528 1 A
16845632 16848528 1 A
16846416 16848528 1 C
16846528 16848528 1 C
16846472 16848528 1 A
16848472 16848528 1 G // вот на этой итерации мы освобождаем память из под end
16848528 -572662307 1 C
16848584 -572662307 1 A
16846680 -572662307 1 G
16847240 -572662307 1 C
16848248 -572662307 1 T
```

ArtyomLobanov replied:
Самый простой вариант исправления – сохранить end->nextElement в отдельную переменную в начале работы функции:
```cpp
void deleteElements(ListElement* start, ListElement* end, List* list)
{
    ListElement* elementAfter = end->nextElement;
    ...
    while (element != elementAfter) {
        ...
    }
}
```

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+        list->tail->nextElement = element;
+        element->previousElement = list->tail;
+        list->tail = element;
+        list->tail->nextElement = NULL;
```
И вот это обнуление указателя также излишне по тем же причинам

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+    }
+}
+
+void findFragment(ListElement* start, ListElement** beginning, ListElement** ending, char* fragment)
```
А как понять, что фрагмент найти не удалось? Давайте какой-нибудь механизм для этого придумаем. Я вижу два варианта:

1.  findFragment возвращает тип bool. Если вернула true – значит поиск удался. Если false – не удался.
2. Если поиск не удался, то и в beginning и в ending записывается значение NULL

ArtyomLobanov replied:
Теоретически, в тестовом задании такая ситуация возникнуть не должна, но это поможет вам обнаруживать ошибки в программе.

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+    if (start == list->head)
+        list->head = element;
+    if (end == list->tail)
+        list->tail = element;
```
Если end был последним элементом списка, то `end->nextElement == NULL`. Из цикла мы выходим только если `end->nextElement == element`, а значит здесь мы запишем в list->tail значение NULL, хотя по хорошему должны были `start->previousElement` (но помните, что память из-под start вы очищаете в цикле)

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+        ListElement* new = makeNewElement(fragment[i]);
+        new->previousElement = current;
+        new->nextElement = current->nextElement;
+        current->nextElement = new;
```
Тут забыли обновить указатель обратный указатель у следующего элемента.
```cpp
current->nextElement->previousElement = new;
```

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,169 @@
+    if (current == NULL) {
+        current = templateEnding->nextElement;
+        if (current == NULL)
+            current = list->head;
```
Какая-то очень сложная конструкция, а главное довольно странная. Если мы заменяем фрагмент в начале списка, то должны добавлять элементы в голову списка, а не после первого элемента. Лучше обучите insertElements корректно работать, если current == NULL. Пусть добавляет элементы в начало списка.

mawekk committed:
⡴⠑⡄⠀⠀⠀⠀⠀⠀⠀ ⣀⣀⣤⣤⣤⣀⡀
⠸⡇⠀⠿⡀⠀⠀⠀⣀⡴⢿⣿⣿⣿⣿⣿⣿⣿⣷⣦⡀
⠀⠀⠀⠀⠑⢄⣠⠾⠁⣀⣄⡈⠙⣿⣿⣿⣿⣿⣿⣿⣿⣆
⠀⠀⠀⠀⢀⡀⠁⠀⠀⠈⠙⠛⠂⠈⣿⣿⣿⣿⣿⠿⡿⢿⣆
⠀⠀⠀⢀⡾⣁⣀⠀⠴⠂⠙⣗⡀⠀⢻⣿⣿⠭⢤⣴⣦⣤⣹⠀⠀⠀⢀⢴⣶⣆
⠀⠀⢀⣾⣿⣿⣿⣷⣮⣽⣾⣿⣥⣴⣿⣿⡿⢂⠔⢚⡿⢿⣿⣦⣴⣾⠸⣼⡿
⠀⢀⡞⠁⠙⠻⠿⠟⠉⠀⠛⢹⣿⣿⣿⣿⣿⣌⢤⣼⣿⣾⣿⡟⠉
⠀⣾⣷⣶⠇⠀⠀⣤⣄⣀⡀⠈⠻⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⠀⠉⠈⠉⠀⠀⢦⡈⢻⣿⣿⣿⣶⣶⣶⣶⣤⣽⡹⣿⣿⣿⣿⡇
⠀⠀⠀⠀⠀⠀⠀⠉⠲⣽⡻⢿⣿⣿⣿⣿⣿⣿⣷⣜⣿⣿⣿⡇
⠀⠀ ⠀⠀⠀⠀⠀⢸⣿⣿⣷⣶⣮⣭⣽⣿⣿⣿⣿⣿⣿⣿⠇
⠀⠀⠀⠀⠀⠀⣀⣀⣈⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠇
⠀⠀⠀⠀⠀⠀⢿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠃

mawekk committed:
:(

mawekk commented:
нашла проблему: на 7 операции (delete) удаляет все, что нужно, кроме первого символа
до этого все работало хорошо
но почему это происходит для меня загадка

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,175 @@
+    ListElement* next = element->nextElement;
+    while (element != endNext) {
+        printf("%c %c\n", element->symbol, next->symbol);
+        element->previousElement->nextElement = next;
```
element->previousElement может быть NULL

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,175 @@
+void insertElements(ListElement* current, char* fragment, List* list)
+{
+    if (!current)
+        current = list->head;
```
Нет, это неправда. Вот у вас есть список
```
        head                     tail
NULL <-> [A] <-> [B] <-> [C] <-> [D] <-> NULL
```
Если вызвать `insertElements(list->head, "EF", list)` или что то же самое `insertElements([A], "EF", list)`, то мы получим такую картину.
```
        head                                       tail
NULL <-> [A] <-> [E]  <-> [F]  <-> [B] <-> [C] <-> [D] <-> NULL
```
А если мы вызываем `insertElements(NULL, "EF", list)`, то хотим получить следующее:
```
        head                                       tail
NULL <-> [E] <-> [F]  <-> [A]  <-> [B] <-> [C] <-> [D] <-> NULL
```

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,175 @@
+        new->previousElement = current;
+        new->nextElement = current->nextElement;
+        current->nextElement = new;
+        current->nextElement->previousElement = new;
```
Эту строчку нужно вставить выше – ведь вы уже записали в `current->nextElement` указатель на new. По сути здесь написано `new->previousElement = new`, что, согласитесь, абсурд.

ArtyomLobanov replied:
И, кстати, учтите, что `current->nextElement` может быть NULL

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,175 @@
+        if (findFragment(startEnding->nextElement, &endBeginning, &endEnding, end))
+            deleteElements(startBeginning, endEnding, list);
+        else
+            printf("ERROR");
```
В качестве дебага – ок, но потом перенесите весь консольный вывод в main. Пусть deleteFragment возвращает bool или код ошибки, чтобы вызывающий его код мог узнать о возникшей проблеме.

ArtyomLobanov commented:
Для сравнения, правильная итоговая последовательность ДНК (т.е. последняя строчка в итоговом файле)
```
GCATGGCAGGGTCATATTGGCGATTAGAGTGCTTCGTTATGCTTCCCCCATCGTCTCGTGTTATCTCCATCCGGAAGTATGGTGTCGCTGGAAAAACTTTGATCAATGGAACTTAACGTAGAGCCTGGGTTGCACTAAAGTAGGATTCGTCGCAAACCCTCAAAACATACGTAGAGCACGCAACCATACCGCGATGGGATCATCGCCGAGCTTTAAATCCATCACCTG
```

mawekk committed:
done 😎

ArtyomLobanov commented on homework_2/task_2.c:
```diff
@@ -0,0 +1,67 @@
+    int size = 0;
+    int numberOfLogs = 0;
+    fscanf(input, "%d", &size);
+    char originalData[size + 1], log[128] = "";
```
Ну я же уже писал про объявление нескольких переменных на одной строке :(

mawekk replied:
ой, мне казалось, я это исправила :(

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,196 @@
+        int i = 0;
+        for (ListElement* element = current; element; element = element->nextElement) {
+            appendElement(string, element->symbol);
+            if (!fragment[i + 1]) {
```
Можно было проще это реализовать:
```cpp
for (ListElement* suffix = start; suffix; suffix = suffix->next) {
    int i = 0;
    for (ListElement* current = suffix; current && current->symbol == fragment[i]; current = current->next) {
        if (!fragment[++i]) { // Строка template кончилась, все предыдущие символы совпали => нашли искомое
            *beginning = current;
            *ending = element;
            return true;
        }
    }
    // Либо элементы списка кончились раньше, чем искомая строка, либо мы нашли несовпадающий символ
}
return false;
```

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,196 @@
+    for (int i = 0; fragment[i]; ++i) {
+        ListElement* new = makeNewElement(fragment[i]);
+        if (!current) {
+            new->nextElement = list->head;
```
Обратную ссылку обновить забыли
`list->head->previousElement = new;`

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,196 @@
+        } else {
+            new->previousElement = current;
+            new->nextElement = current->nextElement;
+            current->nextElement->previousElement = new;
```
current->nextElement может быть NULL, если мы добавляем элементы в конец списка

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,196 @@
+            if (current == list->tail)
+                list->tail = new;
+            if (current == list->head)
+                list->head = new;
```
Почему? Мы же добавляем элементы после current. Если current не NULL, то new никак не может оказаться головой списка

ArtyomLobanov reviewed (approved):
Ура! 👍 
В реализации insertElements есть мелкие баги, которые не влияют на работу на тестовом задании т.к. там ни разу не приходится добавлять элементы в начало. Тем не менее, давайте их исправим. Каждый раз смотря на итерацию цикла или на метод целиком, мысленно в голове прокручивайте инварианты вашей структуры данных и убеждайтесь, что ничего не нарушено. Тогда всякие обратные ссылки теряться не будут.

Пока поставлю 2.5 балла из 3. Исправите insertElements – поставлю полный балл.

mawekk committed:
fix insert + find

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,177 @@
+        ListElement* new = makeNewElement(fragment[i]);
+        if (!current) {
+            new->nextElement = list->head;
+            list->head->previousElement = new;
```
list->head может быть NULL

mawekk committed:
fix insert

mawekk committed:
fix insert (hate styleguide)

ArtyomLobanov commented on library/list.c:
```diff
@@ -0,0 +1,184 @@
+        ListElement* new = makeNewElement(fragment[i]);
+        if (!current) {
+            if (!list->head)
+                appendElement(list, new->symbol);
```
Утечка памяти: вы создали уже выделили память на куче под элемент new, но в этой ветке if-ов никак её не используете и указатель на неё не сохраняете. Либо очищайте память из-под new, если он не пригодился, либо добавляйте элемент в пустой список вручную (благо это делается в две строки), а не вызывайте метод appendElement.
```cpp
list->head = element;
list->tail = element;
```

mawekk committed:
👌🏻

ArtyomLobanov commented:
Ок, зачтено
